File: backend-spring/src/main/java/com/englishlearn/service/DialogueService.java
package com.englishlearn.service;

import com.englishlearn.common.ApiException;
import com.englishlearn.common.JsonUtil;
import com.englishlearn.entity.AssessmentRecord;
import com.englishlearn.entity.ConversationMessage;
import com.englishlearn.entity.ConversationSession;
import com.englishlearn.entity.Scene;
import com.englishlearn.entity.StudyRecord;
import com.englishlearn.entity.User;
import com.englishlearn.llm.EvalResult;
import com.englishlearn.llm.LlmProvider;
import com.englishlearn.llm.Reply;
import com.englishlearn.llm.SummaryResult;
import com.englishlearn.repository.AssessmentRecordRepository;
import com.englishlearn.repository.ConversationMessageRepository;
import com.englishlearn.repository.ConversationSessionRepository;
import com.englishlearn.repository.StudyRecordRepository;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * 对话会话服务（F002/F003）：会话生命周期与消息收发。
 */
@Service
public class DialogueService {

    private static final Logger log = LoggerFactory.getLogger(DialogueService.class);

    private static final int MAX_CONTEXT = 6;

    private final ConversationSessionRepository sessionRepository;
    private final ConversationMessageRepository messageRepository;
    private final AssessmentRecordRepository assessmentRepository;
    private final StudyRecordRepository studyRecordRepository;
    private final LlmProvider llmProvider;
    private final ContentModerationService moderationService;

    /** 模型调用线程池：口译评分在后台异步执行，不阻塞 AI 回复返回 */
    private final ExecutorService llmExecutor = Executors.newFixedThreadPool(4, r -> {
        Thread t = new Thread(r, "llm-caller");
        t.setDaemon(true);
        return t;
    });

    /** 尚未完成的评分任务：结束会话前需等待其落库，避免小结漏算最后一句 */
    private final ConcurrentMap<Integer, Set<CompletableFuture<EvalResult>>> pendingEvals = new ConcurrentHashMap<>();

    public DialogueService(ConversationSessionRepository sessionRepository,
                           ConversationMessageRepository messageRepository,
                           AssessmentRecordRepository assessmentRepository,
                           StudyRecordRepository studyRecordRepository,
                           LlmProvider llmProvider,
                           ContentModerationService moderationService) {
        this.sessionRepository = sessionRepository;
        this.messageRepository = messageRepository;
        this.assessmentRepository = assessmentRepository;
        this.studyRecordRepository = studyRecordRepository;
        this.llmProvider = llmProvider;
        this.moderationService = moderationService;
    }

    @Transactional
    public ConversationSession startSession(User user, Scene scene, String mode) {
        ConversationSession session = new ConversationSession();
        session.userId = user.userId;
        session.scene = scene;
        session.mode = (mode == null || mode.isBlank()) ? "scenario" : mode;
        session.sessionStatus = "ongoing";
        session.startTime = LocalDateTime.now();
        session = sessionRepository.save(session);

        String role = roleOf(scene);
        Reply opening;
        if (scene != null) {
            opening = llmProvider.opening(scene.sceneName, scene.sceneDesc, role);
        } else {
            opening = llmProvider.opening("自由对话", "", "AI 教练");
        }
        ConversationMessage msg = new ConversationMessage();
        msg.sessionId = session.sessionId;
        msg.speaker = "ai";
        msg.contentEn = opening.en();
        msg.contentZh = opening.zh();
        msg.msgTime = LocalDateTime.now();
        messageRepository.save(msg);
        return session;
    }

    public SendResult sendMessage(ConversationSession session, String content) {
        if (!"ongoing".equals(session.sessionStatus)) {
            throw new ApiException(HttpStatus.CONFLICT, "会话已结束，无法继续发送消息");
        }
        Scene scene = session.scene;

        // 内容安全审核：命中不当用语时消息不入库、不进大模型，等价于「撤回」
        ContentModerationService.Decision decision = moderationService.inspect(content);
        if (decision.blocked()) {
            return SendResult.blocked(decision.reason(), decision.tip(),
                    decision.noticeEn(), decision.noticeZh());
        }

        ConversationMessage userMsg = new ConversationMessage();
        userMsg.sessionId = session.sessionId;
        userMsg.speaker = "user";
        userMsg.contentEn = content.strip();
        userMsg.msgTime = LocalDateTime.now();
        userMsg = messageRepository.save(userMsg);

        String role = roleOf(scene);
        List<ConversationMessage> history = recentMessages(session.sessionId);

        // 生成回复在主链路上；口语评分异步进行，不占用用户等待时间
        Reply reply = llmProvider.reply(
                scene != null ? scene.sceneName : "自由对话",
                scene != null ? scene.sceneDesc : "",
                role,
                history,
                content);

        ConversationMessage aiMsg = new ConversationMessage();
        aiMsg.sessionId = session.sessionId;
        aiMsg.speaker = "ai";
        aiMsg.contentEn = reply.en();
        aiMsg.contentZh = reply.zh();
        aiMsg.msgTime = LocalDateTime.now();
        aiMsg = messageRepository.save(aiMsg);

        startAsyncEvaluation(session.sessionId, userMsg.messageId, content);

        return SendResult.of(userMsg, aiMsg, null);
    }

    /**
     * 后台执行口语评分：结果写入 assessment 表，前端通过
     * GET /dialogues/sessions/{id}/messages/{mid}/assessment 轮询获取。
     */
    private void startAsyncEvaluation(Integer sessionId, Integer messageId, String content) {
        CompletableFuture<EvalResult> future = CompletableFuture.supplyAsync(
                () -> llmProvider.evaluate(content), llmExecutor);
        pendingEvals.computeIfAbsent(sessionId, key -> ConcurrentHashMap.newKeySet()).add(future);
        future.whenComplete((eval, error) -> {
            if (eval != null) {
                try {
                    saveAssessment(sessionId, messageId, eval);
                } catch (Exception e) {
                    log.warn("异步口语评分入库失败: {}", e.getMessage());
                }
            } else if (error != null) {
                log.warn("异步口语评分失败: {}", error.getMessage());
            }
            Set<CompletableFuture<EvalResult>> set = pendingEvals.get(sessionId);
            if (set != null) {
                set.remove(future);
                if (set.isEmpty()) pendingEvals.remove(sessionId, set);
            }
        });
    }

    private void saveAssessment(Integer sessionId, Integer messageId, EvalResult eval) {
        AssessmentRecord record = new AssessmentRecord();
        record.sessionId = sessionId;
        record.messageId = messageId;
        record.pronScore = eval.pron;
        record.fluencyScore = eval.fluency;
        record.reactionScore = eval.reaction;
        record.naturalScore = eval.natural;
        record.grammarFeedback = eval.grammarFeedback;
        record.phonemeIssues = eval.phonemeIssues == null || eval.phonemeIssues.isEmpty()
                ? null : JsonUtil.toJson(eval.phonemeIssues);
        record.betterExpression = eval.betterExpression;
        record.assessTime = LocalDateTime.now();
        assessmentRepository.save(record);
    }

    /** 某条用户消息的评分是否已产出 */
    public AssessmentRecord latestAssessment(Integer sessionId, Integer messageId) {
        return assessmentRepository.findFirstBySessionIdAndMessageIdOrderByAssessIdDesc(sessionId, messageId);
    }

    /**
     * 等待本会话尚未完成的评分任务（最长 60 秒），确保结束会话时小结包含最后一句。
     * 注意：必须在事务外调用，避免长时间占用 SQLite 写锁。
     */
    public void awaitPendingEvaluations(Integer sessionId) {
        Set<CompletableFuture<EvalResult>> pending = pendingEvals.remove(sessionId);
        if (pending == null || pending.isEmpty()) {
            return;
        }
        try {
            CompletableFuture.allOf(pending.toArray(new CompletableFuture<?>[0]))
                    .get(60, TimeUnit.SECONDS);
        } catch (Exception ignored) {
            // 超时则以已入库的评分继续，不阻塞会话小结
        }
    }

    @Transactional
    public Map<String, Object> finishSession(ConversationSession session) {
        if ("finished".equals(session.sessionStatus)) {
            Map<String, Object> parsed = JsonUtil.parseMap(session.aiSummary);
            return parsed != null ? parsed : Map.of("summary", Map.of());
        }

        LocalDateTime now = LocalDateTime.now();
        int durationSec = session.startTime != null
                ? Math.max(0, (int) java.time.Duration.between(session.startTime, now).getSeconds())
                : 0;
        session.endTime = now;
        session.durationSec = durationSec;
        session.sessionStatus = "finished";

        List<ConversationMessage> messages = messageRepository.findBySessionIdOrdered(session.sessionId);
        List<AssessmentRecord> assessments = assessmentRepository.findBySessionIdOrderByAssessId(session.sessionId);

        SummaryResult result = llmProvider.summarize(messages, Collections.emptyList());
        if (!assessments.isEmpty()) {
            double mp = assessments.stream().mapToDouble(a -> a.pronScore).average().orElse(0);
            double mf = assessments.stream().mapToDouble(a -> a.fluencyScore).average().orElse(0);
            double mr = assessments.stream().mapToDouble(a -> a.reactionScore).average().orElse(0);
            double mn = assessments.stream().mapToDouble(a -> a.naturalScore).average().orElse(0);
            result.total = round1(mp * 0.35 + mf * 0.25 + mr * 0.2 + mn * 0.2);
            result.pron = round1(mp);
            result.fluency = round1(mf);
            result.reaction = round1(mr);
            result.natural = round1(mn);
        }

        Map<String, Object> dimensions = new LinkedHashMap<>();
        dimensions.put("pron", result.pron);
        dimensions.put("fluency", result.fluency);
        dimensions.put("reaction", result.reaction);
        dimensions.put("natural", result.natural);

        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("sessionId", session.sessionId);
        payload.put("total", result.total);
        payload.put("dimensions", dimensions);
        payload.put("highlights", result.highlights);
        payload.put("improvements", result.improvements);
        payload.put("suggestions", result.suggestions);
        payload.put("corrections", result.corrections);
        payload.put("feedbackText", result.feedbackText);
        payload.put("durationSec", durationSec);
        session.aiSummary = JsonUtil.toJson(payload);
        sessionRepository.save(session);

        int durationMin = Math.max(1, (durationSec + 59) / 60);
        StudyRecord record = new StudyRecord();
        record.userId = session.userId;
        record.sessionId = session.sessionId;
        record.actionType = "scenario";
        record.durationMin = durationMin;
        record.score = result.total;
        record.learnDate = now.toLocalDate();
        studyRecordRepository.save(record);

        return payload;
    }

    public Map<String, Object> getSessionSummary(ConversationSession session) {
        Map<String, Object> parsed = JsonUtil.parseMap(session.aiSummary);
        if (parsed == null) {
            throw new ApiException(HttpStatus.CONFLICT, "会话尚未结束，暂无小结");
        }
        return parsed;
    }

    private List<ConversationMessage> recentMessages(Integer sessionId) {
        List<ConversationMessage> list = new ArrayList<>(messageRepository.findTop6BySessionIdOrderByMessageIdDesc(sessionId));
        Collections.reverse(list);
        return list;
    }

    private String roleOf(Scene scene) {
        if (scene == null) {
            return "";
        }
        Map<String, Object> roleSetting = JsonUtil.parseMap(scene.roleSetting);
        if (roleSetting == null || roleSetting.get("role") == null) {
            return "";
        }
        return String.valueOf(roleSetting.get("role"));
    }

    private static double round1(double v) {
        return Math.round(v * 10.0) / 10.0;
    }

    /**
     * 发送结果。blocked=true 表示命中不当用语被撤回，此时 userMessage / aiMessage 均为 null，
     * 前端应把已上屏的乐观气泡替换为「已撤回」提示，并可展示 noticeEn/noticeZh 的礼貌提醒。
     */
    public record SendResult(ConversationMessage userMessage,
                             ConversationMessage aiMessage,
                             Map<String, Object> liveScores,
                             boolean blocked,
                             String reason,
                             String tip,
                             String noticeEn,
                             String noticeZh) {

        public static SendResult of(ConversationMessage userMessage,
                                    ConversationMessage aiMessage,
                                    Map<String, Object> liveScores) {
            return new SendResult(userMessage, aiMessage, liveScores, false, null, null, null, null);
        }

        public static SendResult blocked(String reason, String tip, String noticeEn, String noticeZh) {
            return new SendResult(null, null, null, true, reason, tip, noticeEn, noticeZh);
        }
    }
}

File: backend/app/services/dialogue_service.py
"""对话会话服务（F002/F003）：会话生命周期与消息收发。"""
import json
from datetime import datetime, timezone

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.models.conversation import AssessmentRecord, ConversationMessage, ConversationSession
from app.models.scene import Scene
from app.models.study import StudyRecord
from app.models.user import User
from app.services.llm import get_llm_provider

_MAX_CONTEXT = 6  # 注入 Provider 的最近上下文条数


def _now() -> datetime:
    return datetime.now(timezone.utc).replace(tzinfo=None)


def _recent_messages(db: Session, session_id: int, limit: int = _MAX_CONTEXT) -> list[ConversationMessage]:
    return list(
        db.scalars(
            select(ConversationMessage)
            .where(ConversationMessage.session_id == session_id)
            .order_by(ConversationMessage.message_id.desc())
            .limit(limit)
        )
    )[::-1]


def start_session(db: Session, user: User, scene: Scene | None, mode: str = "scenario") -> ConversationSession:
    session = ConversationSession(
        user_id=user.user_id,
        scene_id=scene.scene_id if scene else None,
        mode=mode,
        session_status="ongoing",
    )
    db.add(session)
    db.flush()

    provider = get_llm_provider()
    if scene is not None:
        role = (scene.role_setting or {}).get("role", "") if isinstance(scene.role_setting, dict) else ""
        en, zh = provider.opening(scene.scene_name, scene.scene_desc, role)
    else:
        en, zh = provider.opening("自由对话", "", "AI 教练")
    opening = ConversationMessage(
        session_id=session.session_id,
        speaker="ai",
        content_en=en,
        content_zh=zh,
    )
    db.add(opening)
    db.commit()
    db.refresh(session)
    return session


def send_message(
    db: Session, session: ConversationSession, content: str
) -> tuple[ConversationMessage, ConversationMessage, dict]:
    """落库用户消息 → Provider 生成回复并落库 → 启发式评测。返回 (用户消息, AI消息, liveScores)。"""
    if session.session_status != "ongoing":
        raise ValueError("会话已结束，无法继续发送消息")

    scene = session.scene
    user_msg = ConversationMessage(
        session_id=session.session_id, speaker="user", content_en=content.strip()
    )
    db.add(user_msg)
    db.flush()

    provider = get_llm_provider()
    history = _recent_messages(db, session.session_id)
    role = (scene.role_setting or {}).get("role", "") if scene and isinstance(scene.role_setting, dict) else ""
    en, zh = provider.reply(
        scene_name=scene.scene_name if scene else "自由对话",
        scene_desc=scene.scene_desc if scene else "",
        role=role,
        history=history,
        user_input=content,
    )
    ai_msg = ConversationMessage(
        session_id=session.session_id, speaker="ai", content_en=en, content_zh=zh
    )
    db.add(ai_msg)
    db.flush()

    eval_result = provider.evaluate(content)
    record = AssessmentRecord(
        session_id=session.session_id,
        message_id=user_msg.message_id,
        pron_score=eval_result.pron,
        fluency_score=eval_result.fluency,
        reaction_score=eval_result.reaction,
        natural_score=eval_result.natural,
        grammar_feedback=eval_result.grammar_feedback,
        phoneme_issues=eval_result.phoneme_issues,
        better_expression=eval_result.better_expression,
    )
    db.add(record)
    db.commit()

    live_scores = {
        "pron": eval_result.pron,
        "fluency": eval_result.fluency,
        "reaction": eval_result.reaction,
        "natural": eval_result.natural,
    }
    return user_msg, ai_msg, live_scores


def finish_session(db: Session, session: ConversationSession) -> dict:
    """结束会话并生成小结：汇总四维均分、写 ai_summary、沉淀学习记录。"""
    if session.session_status == "finished":
        return json.loads(session.ai_summary) if session.ai_summary else {"summary": {}}

    now = _now()
    duration_sec = max(0, int((now - session.start_time).total_seconds())) if session.start_time else 0
    session.end_time = now
    session.duration_sec = duration_sec
    session.session_status = "finished"

    messages = list(
        db.scalars(
            select(ConversationMessage)
            .where(ConversationMessage.session_id == session.session_id)
            .order_by(ConversationMessage.msg_time)
        )
    )
    assessments = list(
        db.scalars(
            select(AssessmentRecord)
            .where(AssessmentRecord.session_id == session.session_id)
            .order_by(AssessmentRecord.assess_id)
        )
    )
    provider = get_llm_provider()
    result = provider.summarize(messages, [])
    if assessments:
        def mean(key: str) -> float:
            vals = [float(getattr(a, key)) for a in assessments]
            return round(sum(vals) / len(vals), 1)

        result.total = round(mean("pron_score") * 0.35 + mean("fluency_score") * 0.25
                             + mean("reaction_score") * 0.2 + mean("natural_score") * 0.2, 1)
        result.pron, result.fluency = mean("pron_score"), mean("fluency_score")
        result.reaction, result.natural = mean("reaction_score"), mean("natural_score")

    payload = {
        "sessionId": session.session_id,
        "total": result.total,
        "dimensions": {
            "pron": result.pron,
            "fluency": result.fluency,
            "reaction": result.reaction,
            "natural": result.natural,
        },
        "highlights": result.highlights,
        "improvements": result.improvements,
        "suggestions": result.suggestions,
        "corrections": result.corrections,
        "feedbackText": result.feedback_text,
        "durationSec": duration_sec,
    }
    session.ai_summary = json.dumps(payload, ensure_ascii=False)

    # 沉淀学习记录（联动 F007 报表）
    learn_date = now.date()
    duration_min = max(1, (duration_sec + 59) // 60)
    db.add(
        StudyRecord(
            user_id=session.user_id,
            session_id=session.session_id,
            action_type="scenario",
            duration_min=duration_min,
            score=result.total,
            learn_date=learn_date,
        )
    )
    db.commit()
    return payload


def get_session_summary(db: Session, session: ConversationSession) -> dict:
    if not session.ai_summary:
        raise LookupError("会话尚未结束，暂无小结")
    return json.loads(session.ai_summary)
